drivers/i2s/i2s_cavs.c

```diff
@@ -0,0 +1,791 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief I2S bus (SSP) driver for Intel CAVS.
+ *
+ * Limitations:
+ * - DMA is used in simple single block transfer mode (with linked list
+ *   enabled) and "interrupt on full transfer completion" mode.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <misc/__assert.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <dma.h>
+#include <i2s.h>
+#include <soc.h>
+#include "i2s_cavs.h"
+
+#define SYS_LOG_DOMAIN "dev/i2s_cavs"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_I2S_LEVEL
+#include <logging/sys_log.h>
+
+#ifdef CONFIG_DCACHE_WRITEBACK
+#define DCACHE_INVALIDATE(addr, size) \
+		{ dcache_invalidate_region(addr, size); }
+#define DCACHE_CLEAN(addr, size) \
+		{ dcache_writeback_region(addr, size); }
+#else
+#define DCACHE_INVALIDATE(addr, size) \
+		do { } while (0)
+
+#define DCACHE_CLEAN(addr, size) \
+		do { } while (0)
+#endif
+
+#define CAVS_SSP_WORD_SIZE_BITS_MIN    2
+#define CAVS_SSP_WORD_SIZE_BITS_MAX   32
+#define CAVS_SSP_WORD_PER_FRAME_MIN    1
+#define CAVS_SSP_WORD_PER_FRAME_MAX   16
+
+struct queue_item {
+	void *mem_block;
```

Comment:
Any reason you can't use a k_queue for this?  Among other things, I don't see any synchronization of the I2C API here, and a k_queue would give you that for free.

Reply:
Guess a k_queue should work here. I don't need any synchronization though. Will give the struct a try and see if things work fine.